// @stringsync/vexml/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NoopLogger = exports.MemoryLogger = exports.ConsoleLogger = exports.SimpleCursor = exports.MaxMeasureFormatter = exports.MonitoredFormatter = exports.PanoramicFormatter = exports.DefaultFormatter = exports.Score = exports.Renderer = exports.DEFAULT_CONFIG = exports.CONFIG = exports.MXLParser = exports.MusicXMLParser = exports.renderMXL = exports.renderMusicXML = void 0;
var render_1 = require("./render");
Object.defineProperty(exports, "renderMusicXML", { enumerable: true, get: function () { return render_1.renderMusicXML; } });
Object.defineProperty(exports, "renderMXL", { enumerable: true, get: function () { return render_1.renderMXL; } });
__exportStar(require("./elements/eventmappingfactory"), exports);
var parsing_1 = require("./parsing");
Object.defineProperty(exports, "MusicXMLParser", { enumerable: true, get: function () { return parsing_1.MusicXMLParser; } });
Object.defineProperty(exports, "MXLParser", { enumerable: true, get: function () { return parsing_1.MXLParser; } });
var config_1 = require("./config");
Object.defineProperty(exports, "CONFIG", { enumerable: true, get: function () { return config_1.CONFIG; } });
Object.defineProperty(exports, "DEFAULT_CONFIG", { enumerable: true, get: function () { return config_1.DEFAULT_CONFIG; } });
var rendering_1 = require("./rendering");
Object.defineProperty(exports, "Renderer", { enumerable: true, get: function () { return rendering_1.Renderer; } });
var elements_1 = require("./elements");
Object.defineProperty(exports, "Score", { enumerable: true, get: function () { return elements_1.Score; } });
var formatting_1 = require("./formatting");
Object.defineProperty(exports, "DefaultFormatter", { enumerable: true, get: function () { return formatting_1.DefaultFormatter; } });
Object.defineProperty(exports, "PanoramicFormatter", { enumerable: true, get: function () { return formatting_1.PanoramicFormatter; } });
Object.defineProperty(exports, "MonitoredFormatter", { enumerable: true, get: function () { return formatting_1.MonitoredFormatter; } });
Object.defineProperty(exports, "MaxMeasureFormatter", { enumerable: true, get: function () { return formatting_1.MaxMeasureFormatter; } });
var components_1 = require("./components");
Object.defineProperty(exports, "SimpleCursor", { enumerable: true, get: function () { return components_1.SimpleCursor; } });
var debug_1 = require("./debug");
Object.defineProperty(exports, "ConsoleLogger", { enumerable: true, get: function () { return debug_1.ConsoleLogger; } });
Object.defineProperty(exports, "MemoryLogger", { enumerable: true, get: function () { return debug_1.MemoryLogger; } });
Object.defineProperty(exports, "NoopLogger", { enumerable: true, get: function () { return debug_1.NoopLogger; } });
