// @stringsync/vexml/dist/cjs/rendering/clef.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Clef = void 0;
const vexflow = __importStar(require("vexflow"));
const ADDITIONAL_CLEF_WIDTH = 10;
class Clef {
    config;
    log;
    document;
    key;
    constructor(config, log, document, key) {
        this.config = config;
        this.log = log;
        this.document = document;
        this.key = key;
    }
    render() {
        const clef = this.document.getStave(this.key).signature.clef;
        let annotation;
        if (clef.octaveShift) {
            const direction = clef.octaveShift > 0 ? 'va' : 'vb';
            annotation = `8${direction}`;
        }
        const vexflowClef = new vexflow.Clef(clef.sign, 'default', annotation);
        const width = vexflowClef.getWidth() + ADDITIONAL_CLEF_WIDTH;
        return {
            type: 'clef',
            key: this.key,
            width,
            sign: clef.sign,
            vexflowClef,
        };
    }
}
exports.Clef = Clef;
