// @stringsync/vexml/dist/cjs/rendering/octaveshift.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.OctaveShift = void 0;
const vexflow = __importStar(require("vexflow"));
const util = __importStar(require("../util"));
const spatial_1 = require("../spatial");
class OctaveShift {
    config;
    log;
    document;
    key;
    registry;
    constructor(config, log, document, key, registry) {
        this.config = config;
        this.log = log;
        this.document = document;
        this.key = key;
        this.registry = registry;
    }
    render() {
        const octaveShift = this.document.getOctaveShift(this.key);
        const renders = this.registry.get(octaveShift.id).filter((r) => r.type === 'note' || r.type === 'rest');
        const vexflowTextBrackets = this.renderVexflowTextBrackets(renders);
        return {
            type: 'octaveshift',
            key: this.key,
            rect: spatial_1.Rect.empty(),
            vexflowTextBrackets,
        };
    }
    renderVexflowTextBrackets(renders) {
        if (renders.length < 2) {
            this.log.warn('cannot render octave shift with less than 2 notes, skipping', {
                octaveShiftIndex: this.key.octaveShiftIndex,
            });
            return [];
        }
        const vexflowTextBrackets = new Array();
        const systemIndexes = util.unique(renders.map((n) => n.key.systemIndex));
        for (const systemIndex of systemIndexes) {
            const systemNoteRenders = renders.filter((n) => n.key.systemIndex === systemIndex);
            if (systemNoteRenders.length > 1) {
                const vexflowTextBracket = this.renderVexflowTextBracket(systemNoteRenders);
                vexflowTextBrackets.push(vexflowTextBracket);
            }
        }
        return vexflowTextBrackets;
    }
    renderVexflowTextBracket(renders) {
        const octaveShift = this.document.getOctaveShift(this.key);
        const start = renders.at(0).vexflowNote;
        const stop = renders.at(-1).vexflowNote;
        const text = Math.abs(octaveShift.size).toString();
        let position;
        if (octaveShift.size < 0) {
            position = vexflow.TextBracketPosition.TOP;
        }
        else {
            position = vexflow.TextBracketPosition.BOTTOM;
        }
        let superscript;
        if (octaveShift.size < 0) {
            superscript = 'va';
        }
        else {
            superscript = 'mb';
        }
        return new vexflow.TextBracket({
            start,
            stop,
            text,
            superscript,
            position,
        });
    }
}
exports.OctaveShift = OctaveShift;
