// @stringsync/vexml/dist/cjs/rendering/pedal.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Pedal = void 0;
const vexflow = __importStar(require("vexflow"));
const util = __importStar(require("../util"));
const spatial_1 = require("../spatial");
class Pedal {
    config;
    log;
    document;
    key;
    registry;
    constructor(config, log, document, key, registry) {
        this.config = config;
        this.log = log;
        this.document = document;
        this.key = key;
        this.registry = registry;
    }
    render() {
        const pedal = this.document.getPedal(this.key);
        const renders = this.registry.get(pedal.id).filter((r) => r.type === 'note' || r.type === 'rest');
        const vexflowPedalMarkings = this.renderVexflowPedalMarkings(renders);
        let vexflowPedalMarkingType;
        switch (pedal.pedalType) {
            case 'text':
                vexflowPedalMarkingType = vexflow.PedalMarking.type.TEXT;
                break;
            case 'bracket':
                vexflowPedalMarkingType = vexflow.PedalMarking.type.BRACKET;
                break;
            case 'mixed':
                vexflowPedalMarkingType = vexflow.PedalMarking.type.MIXED;
                break;
        }
        for (const vexflowPedalMarking of vexflowPedalMarkings) {
            vexflowPedalMarking.setType(vexflowPedalMarkingType);
        }
        return {
            type: 'pedal',
            key: this.key,
            rect: spatial_1.Rect.empty(),
            vexflowPedalMarkings,
        };
    }
    renderVexflowPedalMarkings(renders) {
        const vexflowPedalMarkings = new Array();
        const systemIndexes = util.unique(renders.map((n) => n.key.systemIndex));
        for (const systemIndex of systemIndexes) {
            const systemScopedRenders = renders.filter((n) => n.key.systemIndex === systemIndex);
            if (systemScopedRenders.length > 1) {
                const vexflowPedalMarking = this.renderSinglePedalMarking(systemScopedRenders);
                vexflowPedalMarkings.push(vexflowPedalMarking);
            }
        }
        return vexflowPedalMarkings;
    }
    renderSinglePedalMarking(renders) {
        const vexflowStaveNotes = new Array();
        for (const noteRender of renders) {
            const vexflowNote = noteRender.vexflowNote;
            if (!(vexflowNote instanceof vexflow.StaveNote)) {
                continue;
            }
            const voiceEntry = this.document.getVoiceEntry(noteRender.key);
            if (voiceEntry.type === 'note' && voiceEntry.pedalMark?.pedalMarkType === 'change') {
                // This is required for vexflow to show pedal changes.
                vexflowStaveNotes.push(vexflowNote, vexflowNote);
            }
            else {
                vexflowStaveNotes.push(vexflowNote);
            }
        }
        vexflowStaveNotes.sort((a, b) => a.getAbsoluteX() - b.getAbsoluteX());
        return new vexflow.PedalMarking(vexflowStaveNotes);
    }
}
exports.Pedal = Pedal;
