// @stringsync/vexml/dist/cjs/rendering/tuplet.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tuplet = void 0;
const vexflow = __importStar(require("vexflow"));
const util = __importStar(require("../util"));
const spatial_1 = require("../spatial");
class Tuplet {
    config;
    log;
    document;
    key;
    registry;
    constructor(config, log, document, key, registry) {
        this.config = config;
        this.log = log;
        this.document = document;
        this.key = key;
        this.registry = registry;
    }
    render() {
        const tuplet = this.document.getTuplet(this.key);
        const voiceEntryRenders = this.registry.get(tuplet.id);
        util.assertDefined(voiceEntryRenders);
        util.assert(voiceEntryRenders.length > 1, 'Tuplet must have more than one voice entry');
        let vexflowTupletLocation;
        switch (tuplet.placement) {
            case 'above':
                vexflowTupletLocation = vexflow.Tuplet.LOCATION_TOP;
                break;
            case 'below':
                vexflowTupletLocation = vexflow.Tuplet.LOCATION_BOTTOM;
                break;
        }
        const vexflowStaveNotes = voiceEntryRenders.map((entry) => entry.vexflowNote);
        const vexflowTuplet = new vexflow.Tuplet(vexflowStaveNotes, {
            location: vexflowTupletLocation,
            ratioed: tuplet.showNumber,
        });
        return {
            type: 'tuplet',
            key: this.key,
            rect: spatial_1.Rect.empty(),
            vexflowTuplet,
        };
    }
}
exports.Tuplet = Tuplet;
