// @stringsync/vexml/dist/cjs/rendering/wedge.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Wedge = void 0;
const vexflow = __importStar(require("vexflow"));
const util = __importStar(require("../util"));
const spatial_1 = require("../spatial");
class Wedge {
    config;
    log;
    document;
    key;
    registry;
    constructor(config, log, document, key, registry) {
        this.config = config;
        this.log = log;
        this.document = document;
        this.key = key;
        this.registry = registry;
    }
    render() {
        const wedge = this.document.getWedge(this.key);
        const renders = this.registry.get(wedge.id).filter((r) => r.type === 'note' || r.type === 'rest');
        const vexflowStaveHairpins = this.renderVexflowStaveHairpins(renders);
        return {
            type: 'wedge',
            key: this.key,
            rect: spatial_1.Rect.empty(),
            vexflowStaveHairpins,
        };
    }
    renderVexflowStaveHairpins(renders) {
        if (renders.length < 2) {
            this.log.warn('cannot render wedge with less than 2 notes, skipping', { wedgeIndex: this.key.wedgeIndex });
            return [];
        }
        const vexflowStaveHairpins = new Array();
        const systemIndexes = util.unique(renders.map((n) => n.key.systemIndex));
        for (const systemIndex of systemIndexes) {
            const systemScopedRenders = renders.filter((n) => n.key.systemIndex === systemIndex);
            if (systemScopedRenders.length > 1) {
                const vexflowStaveHairpin = this.renderSingleStaveHairpin(systemScopedRenders);
                vexflowStaveHairpins.push(vexflowStaveHairpin);
            }
        }
        return vexflowStaveHairpins;
    }
    renderSingleStaveHairpin(renders) {
        const wedge = this.document.getWedge(this.key);
        const firstNote = renders.at(0).vexflowNote;
        const lastNote = renders.at(-1).vexflowNote;
        let vexflowStaveHairpinType;
        switch (wedge.wedgeType) {
            case 'crescendo':
                vexflowStaveHairpinType = vexflow.StaveHairpin.type.CRESC;
                break;
            case 'diminuendo':
                vexflowStaveHairpinType = vexflow.StaveHairpin.type.DECRESC;
                break;
        }
        let vexflowModifierPosition;
        switch (wedge.placement) {
            case 'above':
                vexflowModifierPosition = vexflow.Modifier.Position.ABOVE;
                break;
            case 'below':
                vexflowModifierPosition = vexflow.Modifier.Position.BELOW;
                break;
        }
        return new vexflow.StaveHairpin({ firstNote, lastNote }, vexflowStaveHairpinType).setPosition(vexflowModifierPosition);
    }
}
exports.Wedge = Wedge;
