// @stringsync/vexml/dist/cjs/musicxml/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DYNAMIC_TYPES = exports.BEND_TYPES = exports.HARMONIC_PITCH_TYPES = exports.HARMONIC_TYPES = exports.FERMATA_TYPES = exports.FERMATA_SHAPES = exports.TIED_TYPES = exports.SHOW_TUPLET = exports.PEDAL_TYPES = exports.UP_DOWN_STOP_CONTINUE = exports.WEDGE_TYPES = exports.LINE_TYPES = exports.START_STOP_CONTINUE = exports.OVER_UNDER = exports.ABOVE_BELOW = exports.START_STOP = exports.KEY_MODES = exports.TIME_SYMBOLS = exports.SYLLABIC_TYPES = exports.STAVE_TYPES = exports.BEAM_VALUES = exports.CLEF_SIGNS = exports.START_STOP_DISCONTINUE = exports.BARLINE_LOCATIONS = exports.REPEAT_DIRECTIONS = exports.VERTICAL_DIRECTIONS = exports.BAR_STYLES = exports.NOTEHEADS = exports.ACCIDENTAL_TYPES = exports.NOTE_TYPES = exports.STEMS = void 0;
const util_1 = require("../util");
exports.STEMS = new util_1.Enum(['up', 'down', 'double', 'none']);
exports.NOTE_TYPES = new util_1.Enum([
    '1024th',
    '512th',
    '256th',
    '128th',
    '64th',
    '32nd',
    '16th',
    'eighth',
    'quarter',
    'half',
    'whole',
    'breve',
    'long',
    'maxima',
]);
exports.ACCIDENTAL_TYPES = new util_1.Enum([
    'sharp',
    'natural',
    'flat',
    'double-sharp',
    'sharp-sharp',
    'flat-flat',
    'natural-sharp',
    'natural-flat',
    'quarter-flat',
    'quarter-sharp',
    'three-quarters-flat',
    'three-quarters-sharp',
    'sharp-down',
    'sharp-up',
    'natural-down',
    'natural-up',
    'flat-down',
    'flat-up',
    'double-sharp-down',
    'double-sharp-up',
    'flat-flat-down',
    'flat-flat-up',
    'arrow-down',
    'arrow-up',
    'triple-sharp',
    'triple-flat',
    'slash-quarter-sharp',
    'slash-sharp',
    'slash-flat',
    'double-slash-flat',
    'flat-1',
    'flat-2',
    'flat-3',
    'flat-4',
    'sori',
    'koron',
    'other',
]);
exports.NOTEHEADS = new util_1.Enum([
    'arrow down',
    'arrow up',
    'back slashed',
    'circle dot',
    'circle-x',
    'circled',
    'cluster',
    'cross',
    'diamond',
    'do',
    'fa',
    'fa up',
    'inverted triangle',
    'la',
    'left triangle',
    'mi',
    'none',
    'normal',
    're',
    'rectangle',
    'slash',
    'slashed',
    'so',
    'square',
    'ti',
    'triangle',
    'x',
    'other',
]);
exports.BAR_STYLES = new util_1.Enum([
    'dashed',
    'dotted',
    'heavy',
    'heavy-heavy',
    'heavy-light',
    'light-heavy',
    'light-light',
    'none',
    'regular',
    'short',
    'tick',
]);
exports.VERTICAL_DIRECTIONS = new util_1.Enum(['up', 'down']);
exports.REPEAT_DIRECTIONS = new util_1.Enum(['backward', 'forward']);
exports.BARLINE_LOCATIONS = new util_1.Enum(['right', 'left', 'middle']);
exports.START_STOP_DISCONTINUE = new util_1.Enum(['start', 'stop', 'discontinue']);
exports.CLEF_SIGNS = new util_1.Enum(['G', 'F', 'C', 'percussion', 'TAB', 'jianpu', 'none']);
exports.BEAM_VALUES = new util_1.Enum(['backward hook', 'begin', 'continue', 'end', 'forward hook']);
exports.STAVE_TYPES = new util_1.Enum(['alternate', 'cue', 'editorial', 'ossia', 'regular']);
exports.SYLLABIC_TYPES = new util_1.Enum(['begin', 'end', 'middle', 'single']);
exports.TIME_SYMBOLS = new util_1.Enum([
    'common',
    'cut',
    'dotted-note',
    'normal',
    'note',
    'single-number',
    'hidden',
]);
exports.KEY_MODES = new util_1.Enum([
    'none',
    'major',
    'minor',
    'dorian',
    'phrygian',
    'lydian',
    'mixolydian',
    'aeolian',
    'ionian',
    'locrian',
]);
exports.START_STOP = new util_1.Enum(['start', 'stop']);
exports.ABOVE_BELOW = new util_1.Enum(['above', 'below']);
exports.OVER_UNDER = new util_1.Enum(['over', 'under']);
exports.START_STOP_CONTINUE = new util_1.Enum(['start', 'stop', 'continue']);
exports.LINE_TYPES = new util_1.Enum(['dashed', 'dotted', 'solid', 'wavy']);
exports.WEDGE_TYPES = new util_1.Enum(['crescendo', 'diminuendo', 'stop', 'continue']);
exports.UP_DOWN_STOP_CONTINUE = new util_1.Enum(['up', 'down', 'stop', 'continue']);
exports.PEDAL_TYPES = new util_1.Enum([
    'start',
    'stop',
    'sostenuto',
    'change',
    'continue',
    'discontinue',
    'resume',
]);
exports.SHOW_TUPLET = new util_1.Enum(['actual', 'both', 'none']);
exports.TIED_TYPES = new util_1.Enum(['start', 'stop', 'continue', 'let-ring']);
exports.FERMATA_SHAPES = new util_1.Enum([
    'normal',
    'angled',
    'square',
    'double-angled',
    'double-square',
    'double-dot',
    'half-curve',
    'curlew',
]);
exports.FERMATA_TYPES = new util_1.Enum(['upright', 'inverted']);
exports.HARMONIC_TYPES = new util_1.Enum(['unspecified', 'natural', 'artificial']);
exports.HARMONIC_PITCH_TYPES = new util_1.Enum(['unspecified', 'base', 'touching', 'sounding']);
exports.BEND_TYPES = new util_1.Enum(['normal', 'pre-bend', 'release']);
exports.DYNAMIC_TYPES = new util_1.Enum([
    'p',
    'pp',
    'ppp',
    'pppp',
    'ppppp',
    'pppppp',
    'f',
    'ff',
    'fff',
    'ffff',
    'fffff',
    'ffffff',
    'mp',
    'mf',
    'sf',
    'sfp',
    'sfpp',
    'fp',
    'rf',
    'rfz',
    'sfz',
    'sffz',
    'fz',
    'n',
    'pf',
    'sfzp',
]);
