// @stringsync/vexml/dist/cjs/parsing/musicxml/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DYNAMIC_TYPES = exports.TIED_PHASE = exports.NOTE_DURATION_DENOMINATORS = exports.STEM_DIRECTIONS = exports.NOTEHEADS = exports.CLEF_TYPES = exports.KEY_MODES = void 0;
const util_1 = require("../../util");
exports.KEY_MODES = new util_1.Enum([
    'none',
    'major',
    'minor',
    'dorian',
    'phrygian',
    'lydian',
    'mixolydian',
    'aeolian',
    'ionian',
    'locrian',
]);
exports.CLEF_TYPES = new util_1.Enum([
    'treble',
    'french',
    'subbass',
    'baritone-f',
    'bass',
    'baritone-c',
    'tenor',
    'mezzo-soprano',
    'soprano',
    'alto',
    'percussion',
    'tab',
]);
exports.NOTEHEADS = new util_1.Enum([
    '',
    'D0',
    'D1',
    'D2',
    'D3',
    'T0',
    'T1',
    'T2',
    'T3',
    'X0',
    'X1',
    'X2',
    'X3',
    'S1',
    'S2',
    'R1',
    'R2',
    'DO',
    'RE',
    'MI',
    'FA',
    'FAUP',
    'SO',
    'LA',
    'TI',
    'D',
    'H',
    'N',
    'G',
    'M',
    'X',
    'CX',
    'CI',
    'S',
    'SQ',
    'TU',
    'TD',
    'SF',
    'SB',
]);
exports.STEM_DIRECTIONS = new util_1.Enum(['auto', 'up', 'down', 'none']);
exports.NOTE_DURATION_DENOMINATORS = new util_1.Enum([
    '1024',
    '512',
    '256',
    '128',
    '64',
    '32',
    '16',
    '8',
    '4',
    '2',
    '1',
    '1/2',
]);
exports.TIED_PHASE = new util_1.Enum(['start', 'stop', 'continue']);
exports.DYNAMIC_TYPES = new util_1.Enum([
    'p',
    'pp',
    'ppp',
    'pppp',
    'ppppp',
    'pppppp',
    'f',
    'ff',
    'fff',
    'ffff',
    'fffff',
    'ffffff',
    'mp',
    'mf',
    'sf',
    'sfp',
    'sfpp',
    'fp',
    'rf',
    'rfz',
    'sfz',
    'sffz',
    'fz',
    'n',
    'pf',
    'sfzp',
]);
